sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/IOMetrics.java

```diff
@@ -0,0 +1,130 @@
+  private static final String BACKLOG_ELEMENTS = "backlog_elements";
+  private static final String ELEMENTS_WRITTEN = "elements_written";
+  private static final String BYTES_WRITTEN = "bytes_written";
+
+  private static final Counter ELEMENTS_READ_COUNTER = Metrics.counter(IO_NAMESPACE, ELEMENTS_READ);
+  private static final Counter BYTES_READ_COUNTER = Metrics.counter(IO_NAMESPACE, BYTES_READ);
+  private static final Counter ELEMENTS_WRITTEN_COUNTER =
+      Metrics.counter(IO_NAMESPACE, ELEMENTS_WRITTEN);
+  private static final Counter BYTES_WRITTEN_COUNTER =
+      Metrics.counter(IO_NAMESPACE, BYTES_WRITTEN);
+  private static final Gauge BACKLOG_BYTES_GAUGE = Metrics.gauge(IO_NAMESPACE, BACKLOG_BYTES);
+  private static final Gauge BACKLOG_ELEMENTS_GAUGE = Metrics.gauge(IO_NAMESPACE, BACKLOG_ELEMENTS);
+
+  // ------------------------- Source Metrics -------------------------
+
+  /**
+   * Counter of elements read by a source.
+   */
+  public static Counter elementsRead() {
+    return ELEMENTS_READ_COUNTER;
+  }
+
+  /**
+   * Counter of elements read by a source split.
+   * Only report a fixed set of Split IDs as not to overload metrics backends.
+   */
+  public static Counter elementsReadBySplit(String splitId) {
+    return Metrics.counter(IO_SPLITS_NAMESPACE, renderName(splitId, ELEMENTS_READ));
+  }
+
+  /**
+   * Counter of bytes read by a source.
+   */
+  public static Counter bytesRead() {
+    return BYTES_READ_COUNTER;
+  }
+
+  /**
+   * Counter of bytes read by a source split.
+   * Only report a fixed set of Split IDs as not to overload metrics backends.
+   */
+  public static Counter bytesReadBySplit(String splitId) {
+    return Metrics.counter(IO_SPLITS_NAMESPACE, renderName(splitId, BYTES_READ));
+  }
+
+  /**
+   * Gauge for source backlog in bytes.
+   */
+  public static Gauge backlogBytes() {
+    return BACKLOG_BYTES_GAUGE;
+  }
+
+  /**
+   * Gauge for source split backlog in bytes.
+   * Only report a fixed set of Split IDs as not to overload metrics backends.
+   */
+  public static Gauge backlogBytesOfSplit(String splitId) {
+    return Metrics.gauge(IO_SPLITS_NAMESPACE, renderName(splitId, BACKLOG_BYTES));
+  }
+
+  /**
+   * Gauge for source backlog in elements.
+   */
+  public static Gauge backlogElements() {
+    return BACKLOG_ELEMENTS_GAUGE;
+  }
+
+  /**
+   * Gauge for source split backlog in elements.
+   * Only report a fixed set of Split IDs as not to overload metrics backends.
+   */
+  public static Gauge backlogElementsOfSplit(String splitId) {
+    return Metrics.gauge(IO_SPLITS_NAMESPACE, renderName(splitId, BACKLOG_ELEMENTS));
+  }
+
+  // ------------------------- Sink Metrics -------------------------
+
+  /**
+   * Counter of elements written to a sink.
+   */
+  public static Counter elementsWritten() {
+    return ELEMENTS_WRITTEN_COUNTER;
+  }
+
+  /**
+   * Counter of bytes written to a sink.
+   */
+  public static Counter bytesWritten() {
```

Comment:
sinkBytesWritten

to distinguish it from "bytes written to PCollection" or "bytes written to State"